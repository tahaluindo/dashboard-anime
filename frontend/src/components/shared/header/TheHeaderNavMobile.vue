<template>
  <nav class="flex items-center justify-center text-white">
    <the-button-sandwich
      class="text-white"
      ref="buttonSandwich"
      @click.prevent="swapState"
    />
    <the-float-menu ref="navMenu" @click-outside="close">
      <slot />
    </the-float-menu>
  </nav>
</template>

<script lang="ts">
import { defineComponent, ref, watch } from "vue";
import { useRouter } from "vue-router";
import TheButtonSandwich from "../button/TheButtonSandwich.vue";
import TheFloatMenu from "../menu/TheFloatMenu.vue";

export default defineComponent({
  components: {
    TheButtonSandwich,
    TheFloatMenu,
  },
  setup() {
    const navMenu = ref<InstanceType<typeof TheFloatMenu>>();
    const buttonSandwich = ref<InstanceType<typeof TheButtonSandwich>>();

    const router = useRouter();
    watch(
      () => router.currentRoute.value.path,
      () => {
        close();
      }
    );

    const close = () => {
      navMenu.value?.animationController.close();
      buttonSandwich.value?.animationController.close();
    };

    const swapState = async () => {
      navMenu.value?.animationController.swap();
      buttonSandwich.value?.animationController.swap();
    };

    return {
      navMenu,
      buttonSandwich,
      swapState,
      close,
    };
  },
});
</script>

<style lang="postcss"></style>
