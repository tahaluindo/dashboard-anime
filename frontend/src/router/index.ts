import { createRouter, createWebHistory, RouteRecordRaw } from "vue-router";
import AddUser from "../views/AddUser.vue";
import AllUsers from "../views/AllUsers.vue";
import Home from "../views/Home.vue";
import User from "../views/User.vue";

const routes: Array<RouteRecordRaw> = [
  {
    path: "/",
    name: "Home",
    component: Home,
  },
  {
    path: "/usuarios",
    name: "AllUsers",
    component: AllUsers,
  },
  {
    path: "/usuario/:id",
    component: User,
    name: "User",
    props: true,
  },
  {
    path: "/adicionar-usuario",
    name: "AddUser",
    component: AddUser,
  },
];

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
  scrollBehavior() {
    return window.scrollTo({ top: 0, behavior: "smooth" });
  },
});

export default router;
