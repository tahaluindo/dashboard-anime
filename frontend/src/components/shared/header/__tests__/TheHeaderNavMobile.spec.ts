import { flushPromises, mount } from "@vue/test-utils";
import router from "../../../../router/__mocks__/router.mock";
import { mockWindowVisualViewport } from "../../../../__mocks__/window.mock";
import TheHeaderNavMobile from "../TheHeaderNavMobile.vue";

describe("TheHeaderNavMobile component test", () => {
  it("should render the default component", async () => {
    router.push("/");
    await router.isReady();

    const wrapper = mount(TheHeaderNavMobile, {
      global: {
        stubs: {
          CHeaderNavMobile: true,
        },
        plugins: [router],
      },
    });

    const routerSpy = jest.spyOn(router, "push");
    const buttonSandwich = wrapper.vm.buttonSandwich;
    const navMenu = wrapper.vm.navMenu;

    if (buttonSandwich && navMenu) {
      const navMenuCloseSpy = jest.spyOn(navMenu.animationController, "close");
      const navMenuSwapSpy = jest.spyOn(navMenu.animationController, "swap");
      const buttonSandwichCloseSpy = jest.spyOn(
        buttonSandwich.animationController,
        "close"
      );
      const buttonSandwichSwapSpy = jest.spyOn(
        buttonSandwich.animationController,
        "swap"
      );

      router.push("/contato");
      await router.isReady();
      await wrapper.find(".cursor-pointer").trigger("click.prevent");
      await flushPromises();

      expect(navMenuSwapSpy).toHaveBeenCalled();
      expect(navMenuCloseSpy).toHaveBeenCalled();
      expect(buttonSandwichSwapSpy).toHaveBeenCalled();
      expect(buttonSandwichCloseSpy).toHaveBeenCalled();
    }

    expect(wrapper.html()).toMatchSnapshot();
    expect(routerSpy).toHaveBeenCalled();
  });

  it("should render the default component without refs defined.", async () => {
    mockWindowVisualViewport();

    router.push("/");
    await router.isReady();

    const wrapper = mount(TheHeaderNavMobile, {
      global: {
        plugins: [router],
      },
    });

    const buttonSandwich = wrapper.vm.buttonSandwich;
    const navMenu = wrapper.vm.navMenu;

    if (buttonSandwich && navMenu) {
      const navMenuCloseSpy = jest.spyOn(navMenu.animationController, "close");
      const navMenuSwapSpy = jest.spyOn(navMenu.animationController, "swap");
      const buttonSandwichCloseSpy = jest.spyOn(
        buttonSandwich.animationController,
        "close"
      );
      const buttonSandwichSwapSpy = jest.spyOn(
        buttonSandwich.animationController,
        "swap"
      );

      wrapper.vm.navMenu = undefined;
      wrapper.vm.buttonSandwich = undefined;

      wrapper.vm.close();
      wrapper.vm.swapState();

      expect(navMenuSwapSpy).not.toHaveBeenCalled();
      expect(navMenuCloseSpy).not.toHaveBeenCalled();
      expect(buttonSandwichSwapSpy).not.toHaveBeenCalled();
      expect(buttonSandwichCloseSpy).not.toHaveBeenCalled();
    }

    expect(wrapper.html()).toMatchSnapshot();
  });
});
