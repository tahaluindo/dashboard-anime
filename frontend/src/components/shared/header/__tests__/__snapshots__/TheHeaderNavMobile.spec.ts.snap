// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheHeaderNavMobile component test should render the default component 1`] = `
"<nav class=\\"flex items-center justify-center\\">
  <div class=\\"cursor-pointer mr-4\\">
    <div class=\\"h-1 w-7 mb-1 rounded bg-current\\" style=\\"transform: translateY(8.881784197001252e-16px) ;\\"></div>
    <div class=\\"h-1 w-7 mb-1 rounded bg-current\\" style=\\"opacity: 0.9999999999999999; transform: translateY(0px) ;\\"></div>
    <div class=\\"h-1 w-7 rounded bg-current\\" style=\\"transform: translateY(-8.881784197001252e-16px) ;\\"></div>
  </div>
  <nav class=\\"fixed top-0 left-0 flex flex-col items-start w-3/4 h-screen gap-6 py-5 pt-24 bg-black\\" style=\\"transform: translateX(-99.99999999999999vw) ; opacity: 1;\\"></nav>
</nav>"
`;

exports[`TheHeaderNavMobile component test should render the default component without refs defined. 1`] = `
"<nav class=\\"flex items-center justify-center\\">
  <div class=\\"cursor-pointer mr-4\\">
    <div class=\\"h-1 w-7 mb-1 rounded bg-current\\" style=\\"transform: translateY(8.881784197001252e-16px) ;\\"></div>
    <div class=\\"h-1 w-7 mb-1 rounded bg-current\\" style=\\"opacity: 0.9999999999999999; transform: translateY(0px) ;\\"></div>
    <div class=\\"h-1 w-7 rounded bg-current\\" style=\\"transform: translateY(-8.881784197001252e-16px) ;\\"></div>
  </div>
  <nav class=\\"fixed top-0 left-0 flex flex-col items-start w-3/4 h-screen gap-6 py-5 pt-24 bg-black\\" style=\\"transform: translateX(-99.99999999999999vw) ; opacity: 1;\\"></nav>
</nav>"
`;
